### PR TITLE
[FLINK-25392][table-planner]Support new StatementSet syntax in planner and parser

### PR DESCRIPTION
**Pull request by wenlong88**
## What is the purpose of the change
this pr is part of FLIP-190, aims at adding support  new syntax of statement set proposed in FLIP-190

## Brief change log

  - *introduce new syntax: statement set begin insert into T1 select * from T2; insert into T1 select * from T2; end *
  - *support execute a DML/Query with a EXECUTE key word*
  -  *note: SQL client doesn't support the new syntax in this pr, will be solved as a next step 

## Verifying this change

This change added tests and can be verified as follows:

  - *added new statement set test in both parser and planner*

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (no)
  - The serializers: (no)
  - The runtime per-record code paths (performance sensitive): (no)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: (no)
  - The S3 file system connector: (no)

## Documentation

  - Does this pull request introduce a new feature? (yes)
  - If yes, how is the feature documented? (no yet, will update docs when we complete the work at sql client)

### CODE REVIEW

File: flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/calcite/FlinkCalciteSqlValidator.java
Comment (slinkydeveloper) [inline flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/calcite/FlinkCalciteSqlValidator.java:129]: Why have you removed this code?

File: flink-table/flink-sql-parser/src/main/java/org/apache/flink/sql/parser/dml/SqlExecute.java
Comment (slinkydeveloper) [inline flink-table/flink-sql-parser/src/main/java/org/apache/flink/sql/parser/dml/SqlExecute.java:37]: Can you add a SQL statement example here?

File: flink-table/flink-sql-parser/src/main/java/org/apache/flink/sql/parser/dml/SqlStatementSet.java
Comment (slinkydeveloper) [inline flink-table/flink-sql-parser/src/main/java/org/apache/flink/sql/parser/dml/SqlStatementSet.java:34]: Can you add a SQL statement example here?

File: flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/calcite/FlinkCalciteSqlValidatorTest.java
Comment (slinkydeveloper) [inline flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/calcite/FlinkCalciteSqlValidatorTest.java:53]: Please use assertj and remove `ExpectedException`:

```
assertThatThrownBy(() -> plannerMocks.getParser().parse("EXPLAIN UPSERT INTO t1 VALUES(1)"))
  .isInstanceOf(ValidationException.class)
  .hasMessage("UPSERT INTO statement is not supported. Please use INSERT INTO instead.");
```

File: flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/operations/SqlToOperationConverterTest.java
Comment (slinkydeveloper) [inline flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/operations/SqlToOperationConverterTest.java:1402]: Use assertj `assertThat(operation).isInstanceOf(ExplainOperation)`

File: flink-table/flink-table-planner/src/test/scala/org/apache/flink/table/api/TableEnvironmentTest.scala
Comment (slinkydeveloper) [inline flink-table/flink-table-planner/src/test/scala/org/apache/flink/table/api/TableEnvironmentTest.scala:123]: Use assertj `assertThat().isEqualTo()`

File: flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/api/internal/TableEnvironmentImpl.java
Comment (slinkydeveloper) [inline flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/api/internal/TableEnvironmentImpl.java:747]: The message of this exception doesn't make sense if `operation.size() > 1` but not all operations are `ModifyOperation`. Can you improve the error message in such case?

File: flink-table/flink-table-planner/src/test/resources/org/apache/flink/table/planner/plan/nodes/exec/stream/MatchRecognizeJsonPlanTest_jsonplan/testMatch.out
Comment (slinkydeveloper) [inline flink-table/flink-table-planner/src/test/resources/org/apache/flink/table/planner/plan/nodes/exec/stream/MatchRecognizeJsonPlanTest_jsonplan/testMatch.out:466]: Why this plan change?

File: flink-table/flink-table-planner/src/test/scala/org/apache/flink/table/api/TableEnvironmentTest.scala
Comment (slinkydeveloper) [inline flink-table/flink-table-planner/src/test/scala/org/apache/flink/table/api/TableEnvironmentTest.scala:201]: Can you add one test that shows the output of the explain with more than one insert?

File: flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/calcite/FlinkCalciteSqlValidator.java
Comment (slinkydeveloper) [reply flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/calcite/FlinkCalciteSqlValidator.java:129]: I see the issue. I wonder if the proper nested validation of explain and statement set can happen here as well, rather than on the planner side, as this seems the place where the validation belongs to. But I might be mistaken here.

File: flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/operations/SqlToOperationConverter.java
Comment (slinkydeveloper) [inline flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/operations/SqlToOperationConverter.java:990]: No need to instantiate this list, just assign operations in the if branches with either a singleton list or the list returned by `convertSqlStatementSet`.

File: flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/operations/SqlToOperationConverter.java
Comment (slinkydeveloper) [inline flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/operations/SqlToOperationConverter.java:992]: Inline this

File: flink-table/flink-table-planner/src/main/scala/org/apache/flink/table/planner/calcite/FlinkPlannerImpl.scala
Comment (slinkydeveloper) [inline flink-table/flink-table-planner/src/main/scala/org/apache/flink/table/planner/calcite/FlinkPlannerImpl.scala:167]: On master this is not validated, correct?

**Review by slinkydeveloper — COMMENTED** [review]
Left some general comments. It would be nice if someone with sql-parser templates knowledge double checks the templates part of the PR.

The validation code is a bit confusing to me, as it's scattered through 2/3 different components of the codebase. Not sure how much you can do about it in this PR, but perhaps reverting the change on `FlinkCalciteSqlValidator` and adding the validation for SqlExecute and `SqlStatementSet` in `FlinkCalciteSqlValidator` helps a bit?

File: flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/calcite/FlinkCalciteSqlValidator.java
Comment (wenlong88) [reply flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/calcite/FlinkCalciteSqlValidator.java:129]: I didn't get the confusion " validation code scattered through 2/3 different components of the codebase", I just added the validation in the PlannerImpl. I add cases in some part of  planner tests, because we need to make sure that we really support the new syntax e2e, just like other kind of sql statements. 

I reverted the change you made in FLINK-22942, because I think the change was made in the wrong way, and it is necessary in the validation in StatementSet and Execute, that's why I made the change in this pr:
1.  we should not call validateInsert directly because some metadata used in validation would missed.
2. the former change didn't cover explain
3. in my understanding, we put the validation of flink custom node in the PlannerImpl, such as SqlRichExplain/RichSqlInsert, because calcite validator cannot cover them, and we wan to put all of validation in one place. Another way is put the validation in the SqlRichExplain/RichSqlInsert by rewriting the SqlNode#validate, but it would make the validation scattering in every node.

File: flink-table/flink-table-planner/src/main/scala/org/apache/flink/table/planner/calcite/FlinkPlannerImpl.scala
Comment (wenlong88) [reply flink-table/flink-table-planner/src/main/scala/org/apache/flink/table/planner/calcite/FlinkPlannerImpl.scala:167]: yes, insert is not validated before, that is why the plan of match changed too. 
I am not quite familiar with Match, but according to the code generator, the FINAL function is ignored, so there is no really side effect.

**Comment by wenlong88** [comment]
hi, @slinkydeveloper thanks for the detailed review. I have a question on using Assertj: 
do we need transfer all of the tests into assertj when touching a test file?  Existed tests in the file may cause conflicts in the testing framework, and we need to touch much more tests if doing the migration.

**Comment by slinkydeveloper** [comment]
> do we need transfer all of the tests into assertj when touching a test file?

Yep, it's best effort so you should try to do the best you can, but you're not forced to go too deep ending up changing all the classes of our test infra. Just replacing all the assertions in the file you're touching is fine. Old and new assertions can co-exist.

File: flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/calcite/FlinkCalciteSqlValidator.java
Comment (slinkydeveloper) [reply flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/calcite/FlinkCalciteSqlValidator.java:129]: I think it's all right to keep as you've done, and thanks for fixing FLINK-22942 as well.

**Review by slinkydeveloper — APPROVED** [review]

File: flink-table/flink-table-planner/src/test/scala/org/apache/flink/table/api/TableEnvironmentTest.scala
Comment (twalthr) [inline flink-table/flink-table-planner/src/test/scala/org/apache/flink/table/api/TableEnvironmentTest.scala:213]: nit: invalid indention

**Review by twalthr — CHANGES_REQUESTED** [review]
Thanks for the PR @wenlong88. I added some last feedback. Could you also add some documentation or at least create a followup issue in the FLIP-190 umbrella ticket?

File: flink-table/flink-table-planner/src/test/resources/org/apache/flink/table/planner/plan/nodes/exec/stream/MatchRecognizeJsonPlanTest_jsonplan/testMatch.out
Comment (wenlong88) [reply flink-table/flink-table-planner/src/test/resources/org/apache/flink/table/planner/plan/nodes/exec/stream/MatchRecognizeJsonPlanTest_jsonplan/testMatch.out]: It is initially because we add validation on the query of insert which is ignored before. 
I finally found out that the root cause and fixed in the latest commit: when converting sql node to operation, the query of insert would be validated twice and the validator (in SqlValidatorImpl#navigationInMeasure) rewrite the FINAL call twice.

**Comment by wenlong88** [comment]
hi, @twalthr  thanks for the review, I have update the pr addressing your comment and created FLINK-25600 as a following up subtask. It is almost done in my local repo actually, and I will open the pr after this pr is approved.

File: flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/operations/SqlToOperationConverterTest.java
Comment (twalthr) [inline flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/operations/SqlToOperationConverterTest.java:1497]: Please check https://cwiki.apache.org/confluence/pages/viewpage.action?pageId=191336489#FLIP190:SupportVersionUpgradesforTableAPI&SQLPrograms-EXPLAIN

```
However, the PLAN FOR provides no additional benefit and disables specifying the explain details.

Note: Since "PLAN" should be reserved for the JSON plan, we suggest to deprecate the PLAN FOR syntax.
```

Can we test if the explain details work instead?

File: flink-table/flink-sql-parser/src/main/codegen/includes/parserImpls.ftl
Comment (twalthr) [inline flink-table/flink-sql-parser/src/main/codegen/includes/parserImpls.ftl:1781]: Just to double check: Does this include the `WITH` clause?

https://modern-sql.com/feature/with

**Review by twalthr — CHANGES_REQUESTED** [review]
Thanks for the update @wenlong88. I added my last 2 comments. Should be good in the next iteration.

File: flink-table/flink-sql-parser/src/main/codegen/includes/parserImpls.ftl
Comment (wenlong88) [reply flink-table/flink-sql-parser/src/main/codegen/includes/parserImpls.ftl:1781]: yes, OrderedQueryOrExpr includes with clause. do you have any concern on this? Following is the example:
execute|explain insert into t1 with t2 as (select * from t3) select * from t2;
execute|explain with t2 as (select * from t3) select * from t2;
execute|explain statement set begin insert into t1 with t2 as (select * from t3) select * from t2;

**Review by twalthr — APPROVED** [review]
LGTM

**Pull request closed without merging**
